pkg/workceptor/workceptor.go

```diff
@@ -678,25 +688,91 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
 	return retMap, nil
 }
 
+func (w *Workceptor) attemptRemoteAction(remoteNodeID, remoteUnitID, remoteAction string, unit *workUnit) error {
+	timeout := time.After(300 * time.Second)
+	retryInterval := 500 * time.Millisecond
+	shouldTimeout := func() bool {
+		select {
+		case <-timeout:
+			logger.Error("%s of remote work unit timed out", remoteAction)
+			return true
+		default:
+			time.Sleep(retryInterval)
+			return false
+		}
+	}
+
+	for {
+		conn, _, err := w.connectToRemote(remoteNodeID)
+		if err != nil {
+			logger.Error("Connection failed to %s: %s\n", remoteNodeID, err)
+		}
+
+		if conn != nil {
+			_, err = conn.Write([]byte(fmt.Sprintf("work %s %s\n", remoteAction, remoteUnitID)))
+			if err != nil {
+				logger.Error("Write error sending to %s: %s\n", remoteNodeID, err)
+			} else {
+				conn.Close()
+				break
+			}
+			conn.Close()
+		}
+
+		if shouldTimeout() {
+			return nil
+		}
+	}
+
+	if remoteAction == "cancel" {
+		var state int
+		for {
+			w.activeUnitsLock.RLock()
+			state = unit.status.State
+			w.activeUnitsLock.RUnlock()
+			if state == WorkStateSucceeded || state == WorkStateFailed {
+				return nil
+			}
+
+			if shouldTimeout() {
+				return nil
+			}
+		}
+	}
+	return nil
+}
+
 // CancelUnit cancels a unit, killing it if it is still running
 func (w *Workceptor) CancelUnit(unitID string) error {
 	//TODO: Handle remote cancellation
 	w.scanForUnits()
 	w.activeUnitsLock.RLock()
+	defer w.activeUnitsLock.RUnlock()
 	unit, ok := w.activeUnits[unitID]
 	if !ok {
-		w.activeUnitsLock.RUnlock()
 		return fmt.Errorf("unknown work unit %s", unitID)
 	}
-	w.activeUnitsLock.RUnlock()
 	if unit.worker != nil {
 		err := unit.worker.Cancel()
 		if err != nil {
 			return err
 		}
 	}
-	unit.status.State = WorkStateFailed
-	unit.status.Detail = "Cancelled"
+	unit.status.LocalCancel = true
+	if unit.status.RemoteUnitID != "" {
+		remoteNodeID := unit.status.Node
+		remoteUnitID := unit.status.RemoteUnitID
+		go w.attemptRemoteAction(remoteNodeID, remoteUnitID, "cancel", unit)
```

Comment:
Should we set the local `unit.status.Detail` to "Cancelling" or "Locally Cancelled" or something like that here?